### PR TITLE
beforeLogin Hook

### PR DESCRIPTION
**Pull request by carlosapgomes**
Hello!

I am reopening [PR #4513](https://github.com/parse-community/parse-server/pull/4513) with a ```beforeLogin Hook``` name.
I think it relates to similar PRs: [#4020](https://github.com/parse-community/parse-server/issues/4020) and [#1016](https://github.com/parse-community/parse-server/issues/1016).

In my use case I need a way to add extra validation to the login process. To do that I think it is important to know who the user is but still be able to abort the whole login flow. 

So, biased by my use case, I think that the definition of ```beforeLogin``` is not **before the login process begins**, but **before it is completed**, that is, the hook should be placed in a point on the login flow where the user has already had her credentials verified by parse-server, but before token creation. 

With that in mind, I added code to allow the following snippet to be called from the login process and still be able to throw errors that would abort it:

```
Parse.Cloud.beforeLogin(function (userLoginData) {
 // custom logic here
 if(condition){
           throw new Parse.Error(Parse.Error.OTHER_CAUSE, 'some msg to the client');
 }else{
  return;
 }
});
```
where ```userLoginData``` is an object:
````
{
  'objectId': user.objectId,
  'username': user.username,
  'email': user.email,
  'name':user.name,
  'createdAt':user.createdAt,
  'updatedAt':user.updatedAt,
  'authProvider': '',
  'authData': {}
}
````
```authProvider``` is a string with the provider name (for login with authProvider) or just ```password``` (for login with username/password).
```authData``` is the same authData object available in the authProvider login flow and is an empty object for login with username/password.

I wrote some tests in spec/BeforeLogin.spec.js, but I am pretty sure that they do not cover all possible side effects.

So I would like to ask for opinions and suggestions especially about the choices I made defining ```beforeLogin```, the way it was implemented and about what other tests should I do.

Thanks

### CODE REVIEW

File: src/RestWrite.js [hunk 1]
```diff
@@ -248,7 +249,21 @@ RestWrite.prototype.handleAuthDataValidation = function(authData) {
       throw new Parse.Error(Parse.Error.UNSUPPORTED_SERVICE,
         'This authentication method is unsupported.');
     }
-    return validateAuthData(authData[provider]);
+    return validateAuthData(authData[provider]).then(() => {
+      if (this.response) {
+      // it is a login call
+        runBeforeLoginHandler({
```
Comment (flovilmart) [inline src/RestWrite.js]: As mentioned previously, the object passed to the beforeLogin hook should match the ones provided in other hooks like beforeSave http://parseplatform.org/Parse-SDK-JS/api/v1.11.0/Parse.Cloud.html#.TriggerRequest

The req.object property should be an instance of Parse.User, all the key types should match their proper types (createdAt/updatedAt as dates), authProvider don’t make sense on req.object but may be left on the top level.

We also expect all other keys like installationId, headers etc.. to be set

Comment (carlosapgomes) [reply src/RestWrite.js]: Hi @flovilmart, sorry for the incomplete PR, I misunderstood your message. I am fixing it.
Thanks,

**Comment by flovilmart** [comment]
No need to be sorry! This happen!
Thanks for keeping up with it!

**Comment by omairvaiyani** [comment]
@carlosapgomes May I ask why this PR was closed? Could you summarise what you've achieved and which areas you need help with? I have a significant use-case for this in my application and look forward to seeing this accepted.

**Comment by addisonElliott** [comment]
@omairvaiyani If it's any assistance, this is on my TODO list that I hopefully will get to soon. I have a similar PR with before/after Session triggers that I hope to work on. Maybe even this week or the next if all goes well.

**Comment by omairvaiyani** [comment]
@addisonElliott excellent, I'll watch this PR and help where possible

**Comment by omairvaiyani** [comment]
@carlosapgomes Not a problem - I'm sure @addisonElliott will still find use in your commits

**Comment by haroonKhan-10p** [comment]
@omairvaiyani did you give it a go?
I need this feature too.

**Comment by haroonKhan-10p** [comment]
@flovilmart can you please say whats left in this PR? So, I can finish it asap, and use it in my current project? Its kinda urgent feature.

**Comment by haroonKhan-10p** [comment]
@carlosapgomes please grant me access on your fork, so I can push commits.

**Comment by flovilmart** [comment]
@haroonKhan-10p please open a new pull request.

**Comment by addisonElliott** [comment]
@haroonKhan-10p I would suggest opening a new pull request and if you use code from an existing PR, link to it and credit the author.

As an FYI, I have a similar PR here #4020 that uses a different approach. I would recommend trying to merge and pick and place these two methods to get beforeLogin/afterLogin working.

**Comment by ferrinho** [comment]
Any news on this? This would be a really great feature.

**Comment by NicksonYap** [comment]
+1

**Comment by omairvaiyani** [comment]
Re-attempted in this [PR](https://github.com/parse-community/parse-server/pull/5445)

**Comment by SebC99** [comment]
Hello @omairvaiyani, if I understand well, the #5445 is not the same PR, as it's not allowing you to have the auth providers you were talking about in your first message here.
Have you achieved this in another way?
I would love to see this PR merged too!

**Pull request closed without merging**